### PR TITLE
Add More GNU v2/v3 Tests

### PR DESCRIPTION
**Pull request by brightprogrammer**
Adding more tests.
Tests taken and modified from [here](https://github.com/ianlancetaylor/demangle/blob/master/cases_test.go)
Source test code is licensed under BSD, modified test code is licensed under LGPL-3.0-only

Tests taken for gnu v2 are from [here](https://github.com/abelbriggs1/gnu2-demangler/blob/master/tests/test_demangler.py)
Both source test code and the modified test code is licensed under LGPL v3

Tests taken for gnu v2 are from `libiberty/testsuite/demangle-expected` from [gcc-2.95.3](https://ftp.fu-berlin.de/unix/languages/gcc/old-releases/gcc-2/gcc-2.95.3.tar.bz2) source code are licensed under GPL v2

### CODE REVIEW

**Review by wargio — APPROVED** [review]

**Comment by wargio** [comment]
looks like we do not handle many cases (the copied code is probably old and should be updated?)

**Comment by brightprogrammer** [comment]
Yes, some fixmes lying around. Will fix those...

**Comment by XVilka** [comment]
```
../src/gnu_v2/cplus-dem.c:3097:2: runtime error: null pointer passed as argument 2, which is declared to never be null
    #0 0x7fcc9dd38e7a in remember_Btype ../src/gnu_v2/cplus-dem.c:3097
    #1 0x7fcc9ddf7828 in demangle_fund_type ../src/gnu_v2/cplus-dem.c:2823
    #2 0x7fcc9ddeb620 in do_type ../src/gnu_v2/cplus-dem.c:2622
    #3 0x7fcc9ddf02ed in do_arg ../src/gnu_v2/cplus-dem.c:3014
    #4 0x7fcc9ddf0b8d in demangle_args ../src/gnu_v2/cplus-dem.c:3247
    #5 0x7fcc9ddfdee2 in demangle_signature ../src/gnu_v2/cplus-dem.c:970
    #6 0x7fcc9ddfdee2 in internal_cplus_demangle ../src/gnu_v2/cplus-dem.c:631
    #7 0x7fcc9dde677f in cplus_demangle_v2 ../src/gnu_v2/cplus-dem.c:588
    #8 0x7fcc9dd406dc in libdemangle_handler_cxx ../src/cxx.c:524
    #9 0x55d5ef20a69b in demangle_with_cxx ../test/test_cxx.c:31186
    #10 0x55d5ef20a32d in main ../test/test_cxx.c:31186
    #11 0x7fcc9d029d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    #12 0x7fcc9d029e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
    #13 0x55d5ef20a454 in _start (/home/runner/work/rz-libdemangle/rz-libdemangle/build-asan/test_cxx+0x4c3454)
```

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good now, just please remove formatting changes. That code will be removed once you rewrite it anyway, no need for that, and it creates unnecessary bulk of changes in this PR.

File: src/cxx.c [hunk 1]
```diff
@@ -69,7 +71,7 @@ typedef struct cxx_replace_pair_t {
 		{ "std::regex_search<" t ">", "std::regex_search<" t ", std::allocator<std::" p "csub_match >, std::regex_traits<" t "> >" }, \
 		{ "std::regex_search<" t ">", "std::regex_search<" t " const*, std::allocator<std::" p "csub_match >, " t ", std::regex_traits<" t "> >" }, \
 		{ "std::regex_search<" t ">", "std::regex_search<iterator<" t ">, std::allocator<std::" p "csub_match >, " t ", std::regex_traits<" t "> >" }, \
-	{ "std::regex_search<std::" p "string>", "std::regex_search<std::char_traits<" t ">, std::allocator<" t ">, std::allocator<std::" p "csub_match >, " t ", std::regex_traits<" t "> >" }
+		{ "std::regex_search<std::" p "string>", "std::regex_search<std::char_traits<" t ">, std::allocator<" t ">, std::allocator<std::" p "csub_match >, " t ", std::regex_traits<" t "> >" }
```
Comment (XVilka) [inline src/cxx.c:74]: This particular change please keep though.

Comment (brightprogrammer) [reply src/cxx.c:74]: Ok

File: src/gnu_v2/cplus-dem.c [hunk 2]
```diff
@@ -1422,8 +1423,9 @@ demangle_template(struct work_stuff *work, const char **mangled, string *tname,
 	if (is_java_array) {
 		string_append(tname, "[]");
 	} else {
-		if (tname->p[-1] == '>')
-			string_append(tname, " ");
+    // Test cases failing because of this extra space
```
Comment (XVilka) [inline src/gnu_v2/cplus-dem.c:1426]: Looks like this formatting doesn't match the original one

File: src/gnu_v2/cplus-dem.c [hunk 3]
```diff
@@ -2498,8 +2500,14 @@ do_type(struct work_stuff *work, const char **mangled, string *result)
 				(*mangled)++;
 			break;
 
-		case 'M':
 		case 'O': {
+      (*mangled)++;
```
Comment (XVilka) [inline src/gnu_v2/cplus-dem.c:2504]: Same, please make it match the original formatting.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: src/gnu_v2/cplus-dem.c (hunk 2 above)
Comment (brightprogrammer) [reply src/gnu_v2/cplus-dem.c:1426]: this is a tab to spacing issue I think. This looks fine in my editor
image

Comment (XVilka) [reply src/gnu_v2/cplus-dem.c:1426]: Just install a plugin/command/whatever to optionally show tabs/spaces, like in my NeoVim, for example:

Screenshot 2024-12-22 at 6 54 41 PM

I use this NeoVim config lines for that:
```lua
-- Show tab characters
local tabsshowfn = function()
	vim.o.list = true
	vim.o.listchars = "tab:>-"
end

local tabshidefn = function()
	vim.o.list = false
	vim.o.listchars = ""
end

local tabstogglefn = function()
	if vim.o.list then
		tabshidefn()
	else
		tabsshowfn()
	end
end

vim.keymap.set("n", "<f3>", tabstogglefn)
```

Comment (brightprogrammer) [reply src/gnu_v2/cplus-dem.c:1426]: Ok 👍

File: src/gnu_v2/cplus-dem.c [hunk 4]
```diff
@@ -1075,8 +1075,9 @@ demangle_template_template_parm(struct work_stuff *work, const char **mangled, s
 			need_comma = 1;
 		}
 	}
-	if (tname->p[-1] == '>')
-		string_append(tname, " ");
+  // test cases failing because of this space
```
Comment (XVilka) [inline src/gnu_v2/cplus-dem.c:1078]: You forgot this one

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Review by XVilka — CHANGES_REQUESTED** [review]
Split some tests into a new category with `RZ_DEMANGLE_OPT_BASE`

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@wargio take a look again.

**Review by wargio — APPROVED** [review]
i'm ok with the changes, i just don't see any real upgrade of the old code, so i guess it's ok?

File: src/gnu_v2/cplus-dem.c [hunk 5]
```diff
@@ -2498,8 +2500,14 @@ do_type(struct work_stuff *work, const char **mangled, string *result)
...
 			break;
 
-		case 'M':
 		case 'O': {
+			(*mangled)++;
+			string_prepend(&decl, "&&");
+			if (tk == tk_none)
+				tk = tk_reference;
+			break;
+    }
```
Comment (wargio) [inline src/gnu_v2/cplus-dem.c:2509]: not formatted?

Comment (brightprogrammer) [reply src/gnu_v2/cplus-dem.c:2509]: Ah missed that, sorry 😅

**Comment by XVilka** [comment]
> i'm ok with the changes, i just don't see any real upgrade of the old code, so i guess it's ok?

That's the plan. Cover as much as possible with tests before starting actual rewrite. So that we could be sure the new code works properly.

**Comment by brightprogrammer** [comment]
Yes, and in total, only around 300 tests fail in c++ v3 test set. Some return `NULL` and some are completely wrong. Rewrite will aim to cover all these (and give a more permissive license as well)

**Pull request merged**